### DISCUSSION TITLE
How to generate images in bulk  - performance - web workers

### DISCUSSION THREAD

**Discussion by lh-sn** [fabricjs/fabric.js#9491] [Q&A]
What is the best way to generate images in bulk in the browser. Currently it takes me about 50s for 25 images, so, I have decided to have a try with web workers.

```ts
export class StaticCanvas extends StaticCanvasBase {
  async toBlobURL(options = {} as TDataUrlOptions): Promise<string> {
    const {
      format = 'png' as ImageFormat,
      quality = 1,
      multiplier = 1,
      enableRetinaScaling = false,
    } = options;
    const finalMultiplier =
      multiplier * (enableRetinaScaling ? this.getRetinaScaling() : 1);
    return toBlobURL(
      this.toCanvasElement(
        finalMultiplier,
        options
      ) as unknown as OffscreenCanvas,
      format,
      quality
    );
  }
  constructor(canvasEl: OffscreenCanvas) {
    super(proxyWebAPIs(canvasEl));
  }
}
```

Tried to use webworkers to test it, but I am getting some errors when I try to generate an image:

```ts
import { StaticCanvas, Rect, Image } from 'fabric/offscreen';

const canvas = new StaticCanvas(data.canvas, {
        width: 1200,
        height: 1200,
      });
      const rect = new Rect({
        width: 100,
        height: 100,
        fill: 'red',
        strokeWidth: 0,
      });
      canvas.add(rect);
      const image = await Image.fromURL('', {}, {});

```

I get the following error:

```
Uncaught (in promise) ReferenceError: Image is not defined
```

Any suggestion here ?

**Comment by ShaMan123** [comment]
I understand you tried #8774 
There is no DOM in a web worker and therefore no Image.
Investiage what can be done.

**Comment by asturur** [comment]
50s for 25 images seems like 2 second per render, quite a lot. How does images looks like?

**Reply by lh-sn** [reply]
4 custom fonts, 4 images, 2 vectors.
1. Await Fetch static resources: fonts, images
2. Await StaticCanvas load from JSON 
3. canvas toDataURL

50 times, each with different images and some different fonts
